Never researched option

MaxMallon opened the pull request:
I have added a never-researched category to the inventionslist. It allows to add rides (or scenery groups) into the scenario, to use in constructing the scenario, without them being available to the player when they play the scenario. For instance, a scenario could be that there's a small park that has a single wooden rollercoaster operating "made by a now bankrupt manufactorer", meaning you can't build another wooden coaster.

tupaschoal reviewed (commented):
Can you please rebase instead of merging develop onto your branch? That will make the commits a lot easier to understand and to go through.
(https://github.com/OpenRCT2/OpenRCT2/wiki/Rebase-and-Sync-fork-with-OpenRCT2)

Also there are some changes in `en-GB` which sound misplaced here.

MaxMallon commented:
> Can you please rebase instead of merging develop onto your branch? That will make the commits a lot easier to understand and to go through. (https://github.com/OpenRCT2/OpenRCT2/wiki/Rebase-and-Sync-fork-with-OpenRCT2)

I've done that now

> Also there are some changes in `en-GB` which sound misplaced here.

One edit was indeed misplaced, but the rest is intended

tupaschoal reviewed (commented):
Thanks for rebasing, it really helps. Also since you touched on UI code, can you add screenshots and or a gif of t he new functionality?

MaxMallon commented:
> Thanks for rebasing, it really helps. Also since you touched on UI code, can you add screenshots and or a gif of t he new functionality?

![image](https://user-images.githubusercontent.com/22072563/162380190-1d3c3c96-a772-48b8-9720-9078ade0863e.png)

Changed the "Move all items to top" button to "Move all bottom items to top", as it moves all uninvented items to invented items.
Changed the "Move all items to bottom" button to "Move all top items to bottom", as it moves all invented items to uninvented.
Added "Move all right items to top", which moves all never-invented items to invented items. 

Moved the preview a bit down, and added a scroll for the never invented items. It behaves correctly when rescaling the window.

Items placed in that list will not show up in the rides/themes, and will not be invented.

duncanspumpkin commented on src/openrct2/management/Research.h:
```diff
@@ -46,6 +46,13 @@ enum class ResearchCategory : uint8_t
 
+enum class ResearchStatusType : uint8_t
+{
+    NeverInvented,
```
if you structure this as `Uninvented, Invented, NeverInvented` then import code will be much simpler.

733737 commented:
> > Thanks for rebasing, it really helps. Also since you touched on UI code, can you add screenshots and or a gif of t he new functionality?
> 
> ![image](https://user-images.githubusercontent.com/22072563/162380190-1d3c3c96-a772-48b8-9720-9078ade0863e.png)
> 
> Changed the "Move all items to top" button to "Move all bottom items to top", as it moves all uninvented items to invented items. Changed the "Move all items to bottom" button to "Move all top items to bottom", as it moves all invented items to uninvented. Added "Move all right items to top", which moves all never-invented items to invented items.
> 
> Moved the preview a bit down, and added a scroll for the never invented items. It behaves correctly when rescaling the window.
> 
> Items placed in that list will not show up in the rides/themes, and will not be invented.

I would prefer "Items unavailable during game" ; "Items inventable during game";"All available items Inventable";"All available items Pre-Invented" and "Shuffle available items"

Broxzier commented:
This should be for anything that can be researched, including scenery.

Since this PR has been untouched for a very long time, I'm closing this. The changes do look alright, though it needs to be rewritten with the newer codebase.

MASHzj commented:
Can you explain then what needs to be done in order to get this merged?

MaxMallon commented:
Replying from mail, so not sure if this will even be sent....
I stopped working on it, and my other PRs cause every time I'd update it, people would disagree about something for so long, it needed to be updated again.
If we can come to an agreement on what it must be like, I'll want to work on it again.

Verzonden vanaf Outlook voor Android<https://aka.ms/AAb9ysg>
________________________________
From: MASHzj ***@***.***>
Sent: Tuesday, April 11, 2023 2:46:51 PM
To: OpenRCT2/OpenRCT2 ***@***.***>
Cc: MaxMallon ***@***.***>; Author ***@***.***>
Subject: Re: [OpenRCT2/OpenRCT2] Never researched option (PR #16932)


Can you explain then what needs to be done in order to get this merged?

—
Reply to this email directly, view it on GitHub<https://github.com/OpenRCT2/OpenRCT2/pull/16932#issuecomment-1503264453>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFIMZ44PCODD55W5DGXQSPTXAVHDXANCNFSM5SUHHODA>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

MaxMallon commented:
I did discuss it in the discord back then. Either way, you're saying that all that needs to happen now is updating it again?

Verzonden vanaf Outlook voor Android<https://aka.ms/AAb9ysg>
________________________________
From: Hielke Morsink ***@***.***>
Sent: Tuesday, April 11, 2023 3:15:39 PM
To: OpenRCT2/OpenRCT2 ***@***.***>
Cc: MaxMallon ***@***.***>; Mention ***@***.***>
Subject: Re: [OpenRCT2/OpenRCT2] Never researched option (PR #16932)


@MASHzj<https://github.com/MASHzj> Can you explain then what needs to be done in order to get this merged?

It needs to be rewritten on the latest codebase. The implementation looks alright to me, but I am not very familiar with how this system works.

@MaxMallon<https://github.com/MaxMallon> Features like this are usually best discussed with a developer beforehand. We may have some ideas on how we want to see this implemented, including how the UI should look. On Discord, it's easier to go back and forth on these details, and then when you have created a PR, the reviewing also becomes a lot easier for us.
I'm sorry that none of us have picked this PR up for reviewing. We could've communicated this better with you.

—
Reply to this email directly, view it on GitHub<https://github.com/OpenRCT2/OpenRCT2/pull/16932#issuecomment-1503333180>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFIMZ42OMAGPZAQHXSNQMSTXAVKPXANCNFSM5SUHHODA>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

MASHzj commented:
Honestly it disappoints me what happened here. I commented on this because I'm interested in this feature and to keep this stale PR open. Instead it gets closed immediately with a reason that was not communicated before (or am I missing something?). I also don't see any other objections in this discussion that weren't addressed yet, besides the naming which is quite trivial I think. I'm also not a fan of the current policy of auto-closing PRs, which puts the entire responsibility to the author (I also saw others already saying this).
For some time already I'm considering to contribute, as I have some experience in coding (although not so much in C++). But I don't feel ready for it yet. I'm trying to get familiar with the source code, and how PRs work etc. But seeing things like this happening discourages me to take this effort further.

tupaschoal commented:
> I did discuss it in the discord back then. Either way, you're saying that all that needs to happen now is updating it again?

Seems so, but I'll leave it up for @Broxzier to answer.

> Honestly it disappoints me what happened here. I commented on this because I'm interested in this feature and to keep this stale PR open. Instead it gets closed immediately with a reason that was not communicated before (or am I missing something?).

It was open for an year without activity, so we guessed there was no interest from the OP, even though technically they had implemented it all. We fail too, and Hielke has even already apologized for us not picking this up for review.

> I'm also not a fan of the current policy of auto-closing PRs, which puts the entire responsibility to the author (I also saw others already saying this).

It has only been a couple months since we deployed it, and we've extended the time for the PR to go stale already, because it seemed too short. Bear with us as we also need help to keep things under control.

> But seeing things like this happening discourages me to take this effort further.

What would encourage you to contribute and what would you suggest us do?

MaxMallon commented:
Regarding scenery items, after making this, I'd also made this PR
https://github.com/OpenRCT2/OpenRCT2/pull/16957

We could maybe pick one or the other as it essentially does the same (gameplaywise to the player that is). 

I suggest you have a look at that one too, and then we can decide what does, and what doesn't need to be updated.

MASHzj commented:
@tupaschoal 
Good that the time has been extended, I didn't know that yet. And I understand the mechanism when the author of the PR is asked questions and doesn't respond. But it still feels strange to me if the OP is waiting for feedback or approval from the team. I'm aware that everyone in this project works on it in their free time, and I appreciate that a lot. But that also counts for contributors, and what I'm afraid of is that potential future PRs of mine would close because it might take long to get feedback. And thus the effort I made would feel like a waste of time. Anyway as long as I don't contribute in coding, I try to help on other  things, like I recently answered some feature requests of features that already exist, so those topics can be closed. If there are other things where I could help, you can always let me know. I was for example thinking of adding info to the wiki as I saw that some pages have incomplete information.

@MaxMallon 
I think it's better to continue working on that other PR then, because that one provides more functionalities: there it's possible to mark individual scenery items as hidden, while in this PR it can only be done with scenery groups (if I understand correctly). That's also what I wanted to know with my original question. As I already pointed out in my comment in #16957 , it would be a useful addition to access individual restricted scenery items in RCT1 scenarios.
Also keep in mind (if you haven't already) that restricted rides should not appear in the inventions list anymore. Great that you still want to continue working on this feature!

tupaschoal commented:
Thanks @MASHzj. What you've done helps and updating the wiki is indeed helpful. You can reach out on Discord or on the issue/discussion here if you feel like implementing something, so we can give it a go, that would likely reduce the chance of PR limbo
